nft/ethereum_payment_obligation.go

```diff
@@ -111,111 +115,149 @@ func (s *ethereumPaymentObligation) prepareMintRequest(ctx context.Context, docu
 }
 
 // MintNFT mints an NFT
-func (s *ethereumPaymentObligation) MintNFT(ctx context.Context, documentID []byte, registryAddress, depositAddress string, proofFields []string) (*MintNFTResponse, error) {
+func (s *ethereumPaymentObligation) MintNFT(ctx context.Context, documentID []byte, registryAddress, depositAddress string, proofFields []string) (*MintNFTResponse, chan bool, error) {
 	tc, err := contextutil.Account(ctx)
 	if err != nil {
-		return nil, err
-	}
-
-	requestData, err := s.prepareMintRequest(ctx, documentID, depositAddress, proofFields)
-	if err != nil {
-		return nil, errors.New("failed to prepare mint request: %v", err)
-	}
-
-	opts, err := s.ethClient.GetTxOpts(tc.GetEthereumDefaultAccountName())
-	if err != nil {
-		return nil, err
-	}
-
-	registry := common.HexToAddress(registryAddress)
-	contract, err := s.bindContract(registry, s.ethClient)
-	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	cidBytes, err := tc.GetIdentityID()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	cid, err := identity.ToCentID(cidBytes)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	txID, err := s.sendMintTransaction(cid, contract, opts, requestData, registry, documentID)
+	tokenID := NewTokenID()
+	// Mint NFT within transaction
+	// We use context.Background() for now so that the transaction is only limited by ethereum timeouts
+	txID, done, err := s.txManager.ExecuteWithinTX(context.Background(), cid, uuid.Nil, "Minting NFT",
+		s.minter(ctx, tokenID, documentID, registryAddress, depositAddress, proofFields))
+
 	if err != nil {
-		return nil, errors.New("failed to send transaction: %v", err)
+		return nil, nil, err
 	}
 
 	return &MintNFTResponse{
 		TransactionID: txID.String(),
-		TokenID:       requestData.TokenID.String(),
-	}, nil
+		TokenID:       tokenID.String(),
+	}, done, nil
 }
 
-// OwnerOf returns the owner of the NFT token on ethereum chain
-func (s *ethereumPaymentObligation) OwnerOf(registry common.Address, tokenID []byte) (owner common.Address, err error) {
-	contract, err := s.bindContract(registry, s.ethClient)
-	if err != nil {
-		return owner, errors.New("failed to bind the registry contract: %v", err)
-	}
+func (s *ethereumPaymentObligation) minter(ctx context.Context, tokenID TokenID, documentID []byte, registryAddress, depositAddress string, proofFields []string) func(accountID identity.CentID, txID uuid.UUID, txMan transactions.Manager, errOut chan<- error) {
```

Comment:
I would split the minter funcs maybe into smaller funcs.

Reply:
cyclomatic complexity of this function is more than optimum. Why did we move this here instead of having it part of the documents package? I prefer NFT package doesn't know about the documents package. Mpve this to documents package pls

Reply:
I thing this is refactoring we can consider when we have the entire flow ready, `premature optimisation is the root of all evil` :)

Reply:
I don't consider this to be a premature optimization @vimukthi-git 
I'm talking about breaking down the function to smaller chunks so that its easier to test and understand

Reply:
Agree, just that it seems kind of arbitrary to break it down right now, as only testing we can do is to mock things out individually and see if the function uses mocks correctly. I would love to think about this separately and do a proper test when completing the flow.